### PR TITLE
14: Create Listing

### LINKED ISSUE #14: Create Listing
Request Body
- image
- name
- description
- price
- street
- city
- state
- zipcode

### PR DESCRIPTION
**Pull request by LDomadia**
Notes on testing using Postman:

- First, `npm install` since I added a package (multer) to handle files from POST requests
- Because this post request takes an array of files for the images, the request body must be a "form-data" instead of "json". 
- For the images, the variable is called "imagesArray[]" so that multer recognizes it as an array of files.
- Also, make sure to change it from type "text" to "file".
Screen Shot 2022-12-01 at 10 10 19 PM

- I uploaded the images in Firebase Storage based on the UID (folder name) and the listingId (filename) followed by the file's index in the array. 
- I did it this way just to make things a bit more organized.
Screen Shot 2022-12-01 at 10 14 59 PM

- I also added a listing "title", let me know if I should remove this
- I didn't include "listingId" as a key-value since the doc.id that is generated is already the listingId (similar to how in MongoDB "_id" is the id of the document)

### CODE REVIEW

File: server/src/data/listings.ts [hunk 1]
```diff
@@ -12,6 +14,45 @@ export const getListing = async (listingId: string) => {
   return listing.data();
 };
 
+export const createListing = async (title: String, description: String, price: Number, street: String, city: String, state: String, zip: String, lat: number, lon: number, ownerId: String, imageArray) => {
+  const docRef = await firestore.addDoc(collection, {
+    title: title,
+    description: description,
+    price: price.toFixed(2),
```
Comment (cruiznunez) [inline server/src/data/listings.ts]: by using toFixed, this makes price a string rather than a number in firestore

Comment (LDomadia) [reply server/src/data/listings.ts]: changed line 21 to `price: parseFloat(price.toFixed(2)),`

**Review by uityh — APPROVED** [review]
Nice job!

File: server/src/data/listings.ts [hunk 2]
```diff
@@ -12,6 +14,45 @@ export const getListing = async (listingId: string) => {
   return listing.data();
 };
 
+export const createListing = async (title: String, description: String, price: Number, street: String, city: String, state: String, zip: String, lat: number, lon: number, ownerId: String, imageArray) => {
+  const docRef = await firestore.addDoc(collection, {
+    title: title,
+    description: description,
+    price: parseFloat(price.toFixed(2)),
+    ownerId: ownerId,
```
Comment (uityh) [inline server/src/data/listings.ts:95]: I think we have been calling it owner and not ownerId in firestore.

**Review by uityh — CHANGES_REQUESTED** [review]

**Review by davidtayloryangg — COMMENTED** [review]
Works good job. don't forget to change the data in the database for the stuff you added to test it. Every url for the images should start with the firebasestore and what matt said with ownerId -> owner.

File: server/src/data/listings.ts (hunk 2 above)
Comment (LDomadia) [reply server/src/data/listings.ts:95]: The "bookings" collection has "ownerId", is there anywhere else that says just "owner"? Just want to confirm before I make the change.

Comment (uityh) [reply server/src/data/listings.ts:22]: oh your right, my bad. Just got confused since listings has just owner.

**Review by davidtayloryangg — APPROVED** [review]

**Review by uityh — APPROVED** [review]

**Review by RonnieArvanites — APPROVED** [review]

**Comment by LDomadia** [comment]
- [x] Remove title from collection documents
- [x] Add listingId to collection documents
- [x] Check for duplicate addresses
- [x] At least one image is sent
- [x] Changed "zip" to "zipcode"

File: server/src/data/listings.ts [hunk 3]
```diff
@@ -12,6 +14,47 @@ export const getListing = async (listingId: string) => {
...
+export const createListing = async (description: String, price: Number, street: String, city: String, state: String, zip: String, lat: number, lon: number, ownerId: String, imageArray) => {
+  const docRef = await firestore.addDoc(collection, {
+    description: description,
+    price: parseFloat(price.toFixed(2)),
+    ownerId: ownerId,
+    address: {
+      street: street,
+      city: city,
+      state: state,
+      zip: zip,
```
Comment (cruiznunez) [inline server/src/data/listings.ts]: zip should be a number

Comment (LDomadia) [reply server/src/data/listings.ts]: Zip has to be a string, otherwise 07030 would be stored as 7030

**Pull request merged**